### PR TITLE
Outputs

### PR DESCRIPTION
**Pull request by grahamawelch-zz**
Templates can now have output values that can be consumed using references. Schema validation will no longer validate fields whose value is a reference.

'make test' will now run all of the unit tests under expandybird/expansion.

### CODE REVIEW

**Comment by grahamawelch** [comment]
Fixes #111

**Comment by jackgr** [comment]
Fixes #54.

**Comment by bmelville** [comment]
Does not fix #54

**Comment by jackgr** [comment]
Sorry to be obtuse, but it appears to fix #54 since #117 appears to be a dup of #54. Is that not the case?

**Comment by bmelville** [comment]
#54 is outputs from primitives, #111 is outputs from templates.

**Comment by jackgr** [comment]
So. #54 reads

> As a Kubernetes user, I can write DM templates that bind output parameters from a resource into input parameters in dependent resources.

Is that not outputs from templates?

File: expandybird/expansion/expansion.py [hunk 1]
```diff
@@ -183,6 +194,99 @@ def _ValidateUniqueNames(template_resources, template_name='config'):
     # If this resource doesn't have a name, we will report that error later
 
 
+def IsTemplate(resource_type):
```
Comment (bmelville) [inline expandybird/expansion/expansion.py]: Ville changed the definition of what this means, so you'll need to integrate with how he does it (i.e., it's a template if it exists as an import I think).

File: expandybird/expansion/expansion.py [hunk 2]
```diff
@@ -209,10 +313,19 @@ def ExpandTemplate(resource, imports, env, validate_schema=False):
...
-  if 'path' in imports[source_file] and imports[source_file]['path']:
-      path = imports[source_file]['path']
+  if isinstance(imports[source_file], dict):
+    # This code path assumes a different structure for the 'imports' param.
+    # Map of String (name) to Dict ('path', 'content').
+    #
+    # source_file could be a short version of the template
+    # (say github short name)
+    # so we need to potentially map this into the fully resolvable name.
+    if 'path' in imports[source_file] and imports[source_file]['path']:
```
Comment (bmelville) [inline expandybird/expansion/expansion.py]: I'm confused about how these two clauses differ.

File: expandybird/expansion/expansion_test.py [hunk 3]
```diff
@@ -143,8 +143,11 @@ def testPythonAndJinjaExpansion(self):
     self.assertEquals(result_file, expanded_template)
 
   def testNoImportErrors(self):
+    # TODO(grahamawelch): Ask Ville what whether this test should expect an
```
Comment (bmelville) [inline expandybird/expansion/expansion_test.py]: Can we get this answered before pulling?

File: expandybird/expansion/references.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
+######################################################################
+# Copyright 2015 The Kubernetes Authors All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (bmelville) [inline expandybird/expansion/references.py]: I think these need spaces like https://github.com/kubernetes/deployment-manager/blob/master/common/types.go

File: expandybird/expansion/references_test.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
+######################################################################
+# Copyright 2015 The Kubernetes Authors All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (bmelville) [inline expandybird/expansion/references_test.py]: Same for spaces

File: expandybird/requirements.txt [hunk 6]
```diff
@@ -1,3 +1,6 @@
+--allow-all-external
 pyyaml
 Jinja2
 Jsonschema
+--allow-unverified Jsonpath
+Jsonpath
```
Comment (bmelville) [inline expandybird/requirements.txt]: You will want to make sure you can also build the docker container with this.

File: expandybird/expansion/expansion.py (hunk 1 above)
Comment (grahamawelch) [inline expandybird/expansion/expansion.py]: That doesn't make sense. You can import text files and python libraries that aren't templates.

**Comment by jackgr** [comment]
Need to clear up confusion around issue being addressed. @bmelville please see unanswered question above re: outputs from primitives vs. outputs from templates.

**Comment by jackgr** [comment]
@grahamawelch please reply to googlebot re:cla.

**Comment by grahamawelch** [comment]
I'm trying to sort out all of this merging/rebasing stuff. It will probably end up with me starting over with a new branch/pull request though...

**Comment by grahamawelch** [comment]
This pull request is pretty far out of date at this point so I'll close and issue a new one once I've updated everything.

**Pull request closed without merging**